java/core/src/main/java/org/bondlib/protocol/SimpleBinaryWriter.java

```diff
@@ -48,72 +49,58 @@ public void writeVersion() throws IOException {
 
     @Override
     public void writeFieldOmitted(
-        final BondDataType type, final int id, final FieldMetadata metadata) throws IOException {
+        final BondDataType type, final int id, final Metadata metadata) throws IOException {
         // Simple doesn't support omitting fields, so we need to explicitly write the default values
-        // for optional fields.
-        //
-        // Fields that are = nothing have no default, so it is an unrecoverable error to attempt to
-        // serialize such a field if it is actually set to nothing (e.g., a SomethingObject which is
-        // null).
-
-        if (metadata.isDefaultNothing()) {
-            throw new IllegalArgumentException("can't omit fields with default = nothing in SimpleBinary");
-        }
-
+        // for optional fields. Please note that serialization framework never calls writeFieldOmitted
```

Comment:
Is never supposed to. IIRC, this was a guard against creating garbage data if we incorrectly call this function, which we can't statically prevent. Has something changed?

Reply:
If we have such guards, let's have them in all public methods as I can think of dozens places where unexpected things can happen (e.g. call structEnd without structBegin). This check was not necessary in my opinion as the code guarantees that it won't happen.

An assert could be more appropriate here. I am strictly against replacing would-be-asserts by such exceptions.

---
In reply to: [138762754](https://github.com/Microsoft/bond/pull/611#discussion_r138762754) [](ancestors = 138762754)

Reply:
This is a larger discussion and not relevant to the actual topic of this PR. Could you please revert this change in this PR, and, if you really want to change it, create a new one? If I'm correctly remembering that @chwarr and I added this together, he may have an opinion on it, too.

Reply:
I'll have that discussion with Christopher when he's back. Meanwhile, it bothers me to have code that tests condition that never occurs; again I prefer an assert there. I am reverting this change for now, pending further discussion.